pkg/controllers/cloud/virtualmachine_controller.go

```diff
@@ -33,95 +38,174 @@ const (
+type VirtualMachineManager struct {
 	client.Client
-	Log       logr.Logger
-	Scheme    *runtime.Scheme
-	converter converter.VMConverter
+	log    logr.Logger
+	scheme *runtime.Scheme
 
-	pendingSyncCount int
-	Initialized      bool
+	Inventory inventory.Interface
+	converter converter.VMConverter
+	vmWatcher watch.Interface
 }
 
-// +kubebuilder:rbac:groups=crd.cloud.antrea.io,resources=virtualmachines,verbs=get;list;watch;create;update;patch;delete
-// +kubebuilder:rbac:groups=crd.cloud.antrea.io,resources=virtualmachines/status,verbs=get;update;patch
 // nolint:lll
 
-func (r *VirtualMachineReconciler) Reconcile(_ context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = r.Log.WithValues("Virtualmachine", req.NamespacedName)
-
-	if !r.Initialized {
-		if err := GetControllerSyncStatusInstance().waitTillControllerIsInitialized(&r.Initialized, initTimeout, ControllerTypeVM); err != nil {
-			return ctrl.Result{}, err
-		}
-	}
-
-	virtualMachine := cloudv1alpha1.VirtualMachine{}
-	if err := r.Get(context.TODO(), req.NamespacedName, &virtualMachine); err != nil {
-		if !errors.IsNotFound(err) {
-			r.Log.Error(err, "failed to get VM crd", "vm", req.NamespacedName)
-			return ctrl.Result{}, err
-		}
-		// Fall through if owner is deleted.
-		r.Log.Info("Received request", "vm", req.NamespacedName, "operation", "delete")
-		accessor, _ := meta.Accessor(&virtualMachine)
-		accessor.SetName(req.Name)
-		accessor.SetNamespace(req.Namespace)
-	} else {
-		r.Log.Info("Received request", "vm", req.NamespacedName, "operation", "create/update")
-	}
+func init() {
+	vmManager = &VirtualMachineManager{}
+}
 
-	r.converter.Ch <- virtualMachine
-	r.updatePendingSyncCountAndStatus()
-	return ctrl.Result{}, nil
+// GetVirtualMachineManagerInstance gets the vmManager instance.
+func GetVirtualMachineManagerInstance() *VirtualMachineManager {
+	return vmManager
 }
 
-// SetupWithManager registers VirtualMachineReconciler with manager.
-// It also sets up converter channels to forward VM events.
-func (r *VirtualMachineReconciler) SetupWithManager(mgr ctrl.Manager) error {
+// Configure configures the client, scheme, log and converter.
+func (r *VirtualMachineManager) Configure(client client.Client, scheme *runtime.Scheme,
+	inventory inventory.Interface) *VirtualMachineManager {
+	r.Client = client
+	r.scheme = scheme
+	r.Inventory = inventory
+	r.log = logging.GetLogger("controllers").WithName("VirtualMachine")
 	r.converter = converter.VMConverter{
 		Client: r.Client,
 		Log:    logging.GetLogger("converter").WithName("VMConverter"),
-		Ch:     make(chan cloudv1alpha1.VirtualMachine, converterChannelBuffer),
-		Scheme: r.Scheme,
+		Ch:     make(chan watch.Event, converterChannelBuffer),
+		Scheme: r.scheme,
+	}
+	return r
+}
+
+// Start performs the initialization of the controller.
+// VM controller is said to be initialized and synced when all the VM and EE
+// CRs are reconciled.
+func (r *VirtualMachineManager) Start() error {
```

Comment:
Dont fire goroutine for start in main

Reply:
Changed it. main will call a function to configure VMController, which in turn will start a goroutine.